superset/views/core.py

```diff
@@ -1238,9 +1239,12 @@ def explore_json(self, datasource_type=None, datasource_id=None):
         requests that GETs or POSTs a form_data.
 
         `self.generate_json` receives this input and returns different
-        payloads based on the request args in the first block
+        payloads based on the request args in the first block"""
 
-        TODO: break into one endpoint for each return shape"""
+        if request.method != 'POST' and not is_feature_enabled('CLIENT_CACHE'):
```

Comment:
I think we should avoid coupling the two questions of:
- is CLIENT_CACHE enabled and
- does explore_json support GET

The GET method is an established part of the API here, so its removal constitutes a breaking/non-back-compat change and should be evaluated independently of the cache feature and should include a deprecation process IMO.  I also think the `@expose` configuration should match the behavior.

Consider:
- _deprecating_ GET for all users
- adding a feature flag to turn off GET immediately
- make the `@expose` call `methods` param use a computed value that is something along the lines of the func below...

```
if is_feature_enabled('CLIENT_CACHE') and is_feature_enabled('DISABLE_EXPLORE_JSON_GET'):
  raise Exception('can't have both of these....')
elif is_feature_enabled('DISABLE_EXPLORE_JSON_GET'):
  return ['POST']
else:
  return ['POST', 'GET']
```

Reply:
Also just to be clear, my concern above may not be well-founded--maybe it doesn't need to be treated as a public API--but I did want to raise it for discussion. (Probably good to make the expose params match behavior in either case?)

Reply:
I agree, it's not good to mix CLIENT_CACHE feature flag with GET/POST for explore_json endpoint. So i added another flag, `ENABLE_EXPLORE_JSON_CSRF_PROTECTION`, and set default value be `false`. In default case `explore_json` accept both GET and POST request.

Only when `ENABLE_EXPLORE_JSON_CSRF_PROTECTION` is set to `True`, explore_json will reject GET request method with `405` status code.